### PR TITLE
Improve statistics

### PR DESCRIPTION
**Pull request by p-glaum**
Closes # (if applicable).

## Changes proposed in this Pull Request

Does the aggregation to the wanted grouper format while having the most details. Therefore, one can retrieve the supply for the correct location and can group it for the countries. This is still compatible with the former grouper functions and should be completely backwards compatible.
Furthermore, now one can similarly to `supply` and `withdrawal` split the revenue into `input_cost` and `output_revenue`

## Checklist

- [ ] Improve documentation and examples
- [ ] Code changes are sufficiently documented; i.e. new functions contain docstrings and further explanations may be given in `doc`.
- [ ] Unit tests for new features were added (if applicable).
- [x] Newly introduced dependencies are added to `environment.yaml`, `environment_docs.yaml` and `setup.py` (if applicable).
- [ ] A note for the release notes `doc/release_notes.rst` of the upcoming release is included.
- [x] I consent to the release of this PR's code under the MIT license.

### CODE REVIEW

File: pypsa/statistics.py [hunk 1]
```diff
@@ -127,24 +161,29 @@ def port_mask(n, c, port="", bus_carrier=None):
     if bus_carrier is None:
         mask = n.df(c)[f"bus{port}"] != ""
     elif isinstance(bus_carrier, str):
-        mask = n.df(c)[f"bus{port}"].map(n.buses.carrier).eq(bus_carrier)
+        mask = (
+            n.df(c)[f"bus{port}"]
+            .map(n.buses.carrier)
+            .str.contains(bus_carrier)
```
Comment (FabianHofmann) [inline pypsa/statistics.py:167]: could we address the "string contains" feature later?

**Review by FabianHofmann — COMMENTED** [review]
looks nice @p-glaum, ping me when I should review in detail

File: pypsa/statistics.py (hunk 1 above)
Comment (p-glaum) [reply pypsa/statistics.py]: I already made an improvement there, but we can discuss it as soon as I am finished, and it is ready for review.

File: pypsa/statistics.py [hunk 2]
```diff
@@ -396,24 +428,33 @@ def capex(
...
-                    for port in get_ports(n, c)
-                ]
-                mask = reduce(np.logical_or, masks)
-                col = col[mask.astype(bool)]
-            return col
+            if bus_carrier is None:
+                grouper = get_grouper(n, c, groupby=groupby, nice_names=nice_names)
+                ds = col.groupby(grouper).agg(aggregate_groups)
+                return ds
+            else:
```
Comment (FabianHofmann) [inline pypsa/statistics.py:435]: I would leave the costs mostly untouched. Iterating and summing over different ports will lead to double counting of costs if more than one bus-carrier is selected. Think about a bus_carrier = ["gas", "electricity"], then the cost of components which are attached to both carriers at the same time are summed.

File: pypsa/statistics.py [hunk 3]
```diff
@@ -695,13 +809,19 @@ def supply(
         def func(n, c):
             sign = -1.0 if c in n.branch_components else n.df(c).get("sign", 1.0)
             ports = get_ports(n, c)
-            p = pd.DataFrame(0, index=n.snapshots, columns=n.df(c).index)
+            p = list()
```
Comment (FabianHofmann) [inline pypsa/statistics.py:812]:
```suggestion
            p = []
```

File: pypsa/statistics.py [hunk 4]
```diff
@@ -695,13 +809,19 @@ def supply(
...
             for port in ports:
                 mask = port_mask(n, c, port=port, bus_carrier=bus_carrier)
                 df = n.pnl(c)[f"p{port}"]
-                df = sign * mask * df.reindex(columns=p.columns, fill_value=0)
+                df = sign * df.reindex(columns=mask.index, fill_value=0)
```
Comment (FabianHofmann) [inline pypsa/statistics.py:816]: the mask is now taken only for reindexing, but not used for multiplication anymore. Is that intended?

**Comment by p-glaum** [comment]
Closed in favor of #868

**Comment by p-glaum** [comment]
Closed in favor of #868

**Pull request closed without merging**